src/serial/MidiSessionALSA.cc

```diff
@@ -166,13 +174,253 @@ void MidiOutALSA::serialize(Archive& /*ar*/, unsigned /*version*/)
+	std::string desc;
+	bool connected = false;
+	bool stop = false;
+	cb_queue<uint8_t> queue;
+	std::mutex mutex; // to protect queue
+};
+
+MidiInALSA::MidiInALSA(
+		EventDistributor& eventDistributor_,
+		Scheduler& scheduler_,
+		snd_seq_t& seq_,
+		snd_seq_client_info_t& cinfo, snd_seq_port_info_t& pinfo)
+	: eventDistributor(eventDistributor_)
+	, scheduler(scheduler_)
+	, seq(seq_)
+	, srcClient(snd_seq_port_info_get_client(&pinfo))
+	, srcPort(snd_seq_port_info_get_port(&pinfo))
+	, name(snd_seq_client_info_get_name(&cinfo))
+	, desc(snd_seq_port_info_get_name(&pinfo))
+{
+	eventDistributor.registerEventListener(EventType::MIDI_IN_ALSA, *this);
+	if ((snd_seq_port_info_get_capability(&pinfo) & (SND_SEQ_PORT_CAP_WRITE | SND_SEQ_PORT_CAP_SUBS_WRITE))) {
+		name.append(" (input)");
+	}
+}
+
+MidiInALSA::~MidiInALSA()
+{
+	if (connected) {
+		disconnect();
+	}
+	eventDistributor.unregisterEventListener(EventType::MIDI_IN_ALSA, *this);
+}
+
+void MidiInALSA::plugHelper(Connector& connector_, EmuTime::param /*time*/)
+{
+	auto& midiConnector = checked_cast<MidiInConnector&>(connector_);
+	midiConnector.setDataBits(SerialDataInterface::DATA_8); // 8 data bits
+	midiConnector.setStopBits(SerialDataInterface::STOP_1); // 1 stop bit
+	midiConnector.setParityBit(false, SerialDataInterface::EVEN); // no parity
+
+	setConnector(&midiConnector); // base class will do this in a moment,
+	                              // but thread already needs it
+	connect();
+}
+
+void MidiInALSA::unplugHelper(EmuTime::param /*time*/)
+{
+	if (connected) {
+		disconnect();
+	}
+}
+
+void MidiInALSA::connect()
+{
+	destinationPort = snd_seq_create_simple_port(
+		&seq, "MIDI in pluggable",
+		SND_SEQ_PORT_CAP_WRITE | SND_SEQ_PORT_CAP_SUBS_WRITE, SND_SEQ_PORT_TYPE_MIDI_GENERIC);
+	if (destinationPort < 0) {
+		throw PlugException(
+			"Failed to create ALSA port: ", snd_strerror(destinationPort));
+	}
+
+	int err = snd_seq_connect_from(&seq, destinationPort, srcClient, srcPort);
+	if (err) {
+		snd_seq_delete_simple_port(&seq, destinationPort);
+		throw PlugException(
+			"Failed to connect to ALSA port "
+			"(", srcClient, ':', srcPort, ")"
+			": ", snd_strerror(err));
+	}
+
+	snd_midi_event_new(MidiOutDevice::MAX_MESSAGE_SIZE, &event_parser);
+	snd_midi_event_no_status(event_parser, 1);
+
+	connected = true;
+	stop = false;
+
+	thread = std::thread([this]() { run(); });
+}
+
+void MidiInALSA::disconnect()
+{
+	stop = true;
+	thread.join();
+
+	snd_midi_event_free(event_parser);
+	snd_seq_disconnect_from(&seq, destinationPort, srcClient, srcPort);
+	snd_seq_delete_simple_port(&seq, destinationPort);
+
+	connected = false;
+}
+
+void MidiInALSA::run()
+{
+	assert(isPluggedIn());
+
+	auto npfd = snd_seq_poll_descriptors_count(&seq, POLLIN);
+	std::vector<struct pollfd> pfd(npfd);
+	snd_seq_poll_descriptors(&seq, pfd.data(), npfd, POLLIN);
+
+	while (true) {
+		if (poll(pfd.data(), npfd, 1000) > 0) {
+			do {
+				snd_seq_event_t *ev = NULL;
+				long err, size;
```

Comment:
Detail (coding style).
I personally try to avoid forward declaring variables and leaving them uninitialized. In this case it's easy to avoid:

Write the if below like:
`if (auto err = snd_seq_event_input(...); err < 0) { ...`
And further down:
`if (auto size = snd_midi_event_decode(...); size < 0) { ...`

Bonus:
* This limits the scope of the `err` and `size` variables.
* It allows to use `auto` to deduce the type. Direct use of the type `long` is problematic in portable code (linux/windows, 32bit/64bit) (i know that this code is not supposed to compile on windows, but it's a good habit to avoid `long` anyway). And in this case we don't really care about the exact type of these variable, we only care that it's the same type as is returned from these functions, that's a perfect fit for `auto`.

Reply:
Done except for `size`, since that one needs to be accessed in the later iteration loop.